openforcefield/utils/callback.py

```diff
@@ -0,0 +1,198 @@
+
+import functools
+import inspect
+
+
+# =====================================================================
+# CALLBACKABLE CLASSES
+# =====================================================================
+
+def callback_method(func=None, events=()):
+    """Decorator used to mark a method as callbackable.
+
+    This decorator is designed to be used by ``Callbackable`` children
+    classes. Usage examples can be found in the documentation of ``Callbackable``.
+
+    Parameters
+    ----------
+    func : MethodType
+        The instance method to decorate. Static and class methods are
+        currently not supported.
+    events : Iterable[str]
+        The name of the callback events associated to this method.
+
+    See Also
+    --------
+    Callbackable
+
+    """
+
+    if func is None and len(events) > 0:
+        # Decorator called with @callback_method(events=[...]).
+        return functools.partial(callback_method, events=events)
+
+    @functools.wraps(func)
+    def callbacked_func(self, *args, **kwargs):
+        # Store the returned value.
+        returned_value = func(self, *args, **kwargs)
+        # Invoke callbacks that must be run after the function is executed.
+        self._raise_callback_events(func.__name__, *args, **kwargs)
+        return returned_value
+
+    # Assign an attribute so that we can distinguish between
+    # callbackable and not callbackable methods.
+    callbacked_func._callback_events = set(events)
+
+    return callbacked_func
+
+
+class CallbackRegistrationError(TypeError):
+    """Error raised when callback registration fails."""
+    pass
+
+
+class Callbackable:
+    """A base class for registering and handling callbacks.
+
+    The class allows easy implementation of classes that support the
+    registration of callback function when some instance methods are
+    called. Callback registered to static and class methods are currently
+    not supported.
+
+    The class allows callbacks to be registered to single methods or to
+    an event, which essentially provides a way to register the same callback
+    to multiple methods with a single line.
+
+    Callbacks are called after the function has been executed, and they
+    must have the signature
+
+        callback(callbackable, function_name, *args, **kwargs)
+
+    where ``callbackable`` is the ``Callbackable`` instance that raised
+    the event, ``function_name`` is the name of the function of ``callbackable``
+    that was called, and ``*args`` and ``**kwargs`` are the arguments that
+    were passed to the function.
+
+    Examples
+    --------
+
+    We define a callbackable list that allows registering callback
+    functions to methods that change the list. The first four methods
+    summarize the different ways in which the ``callback_method``
+    decorator can be used to mark a method for callback registration.
+
+    >>> class CallbackableList(Callbackable, list):
```

Comment:
Super clear docstring. This was really helpful in making me understand what this does.